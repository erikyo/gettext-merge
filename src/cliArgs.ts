import yargs from 'yargs'
import { hideBin } from 'yargs/helpers'

export const argv = yargs(hideBin(process.argv))
	.option('in', {
		alias: 'i',
		describe: 'Input .pot files to merge (2 files only)',
		type: 'array',
		demandOption: true,
		coerce: (files: string[]) => {
			if (files.length <= 1) {
				throw new Error('Exactly two or more input files are required.')
			}
			return files
		},
	})
	.option('out', {
		alias: 'o',
		describe: 'Output path for the merged .pot file',
		type: 'string',
		demandOption: true,
	})
	.parse()
