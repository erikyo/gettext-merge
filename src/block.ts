import { mergeComments } from './merge'
import { GetTextComment } from './types'
import { stripQuotes } from './utils'

/**
 * This class represents a single block of PO file.
 */
export class Block {
	comments?: GetTextComment // #| Previous untranslated string
	msgid: string // "%s example"
	msgstr: string[] // ["% esempio", "%s esempi"],
	msgid_plural?: string // "%s examples"
	msgctxt?: string // context

	/**
	 * Constructor for initializing the message properties from the given lines.
	 *
	 * @param {string[]} lines - The array of strings containing the message lines.
	 */
	constructor(lines: string[] | Block) {
		if (lines instanceof Block) {
			this.msgid = lines.msgid
			this.msgid_plural = lines.msgid_plural
			this.msgstr = lines.msgstr
			this.msgctxt = lines.msgctxt
			this.comments = lines.comments
		} else {
			this.msgid = this.parseForLine(lines, /^msgid/, true) as string
			this.msgid_plural = this.parseForLine(lines, /^msgid_plural/, true) as string
			this.msgstr = this.parseForLine(lines, /^msgstr/, false) as string[]
			this.msgctxt = this.parseForLine(lines, /^msgctxt/, true) as string
			this.comments = {
				extracted: this.parseForLine(lines, /^#\./, true) as string,
				reference: this.parseForLine(lines, /^#:/, false) as string[],
				flag: this.parseForLine(lines, /^#,/, true) as string,
				previous: this.parseForLine(lines, /^#\|/, false) as string[],
				translator: this.parseForLine(
					lines,
					/^#(?!\.|:|,|\|)\s?/, // all # that is not #. #: #, or #|
					true
				) as string,
			}
		}
	}

	/**
	 * Parses the given array of strings for a specific id and returns the matching
	 * string or array of strings.
	 *
	 * @param {string[]} lines - the array of strings to parse
	 * @param {string} id - the id to search for
	 * @param {boolean} single - indicates whether to return a single string or an array
	 * @return {string|string[]} the matching string or array of strings
	 */
	private parseForLine(
		lines: string[],
		id: RegExp,
		single: boolean = false
	): string | string[] {
		const res: string[] = []

		let startParse = false

		for (let i = 0; i < lines.length; i++) {
			if (id.test(lines[i]) || (startParse && lines[i].startsWith('"'))) {
				res.push(lines[i]) // Remove quotes at the start and end.
				startParse = true
			} else if (startParse) {
				break // Stop parsing if we have started and now reached a line that doesn't start with a quote
			}
		}

		return single ? res.join('').replace(/""/g, '') : res // Join the array into a single string if single is true.
	}

	/**
	 * Converts the object to a string representation.
	 *
	 * @return {string} The string representation of the object.
	 */
	toStr(): string {
		const { comments, msgid, msgid_plural, msgstr, msgctxt } = this
		const res = [
			comments?.translator,
			comments?.extracted,
			comments?.reference?.join('\n'),
			comments?.flag,
			comments?.previous?.join('\n'),
			msgctxt,
			msgid,
			msgid_plural,
			msgstr.join('\n'),
		]
			.filter(Boolean)
			.filter((i) => i?.length)

		return res.join('\n')
	}

	/**
	 * Generates a hash value for the concatenation of msgctxt, msgid, and msgid_plural.
	 *
	 * @return {number} the hash value generated
	 */
	hash(): number {
		const strToHash = this?.msgctxt + this?.msgid + this?.msgid_plural
		let hash = 0
		for (let i = 0; i < strToHash.length; i++) {
			const chr = strToHash.charCodeAt(i)
			hash = ((hash << 5) - hash + chr) | 0
		}
		return hash
	}

	/**
	 * Merges the other block with the current block.
	 */
	merge(other: Block) {
		if (this.msgid === other.msgid) {
			this.msgid_plural = this.msgid_plural ?? other.msgid_plural
			this.msgctxt = this.msgctxt ?? other.msgctxt
			this.msgstr = this.msgstr ?? other.msgstr
			this.comments = mergeComments(this.comments, other.comments)
		}
	}
}
