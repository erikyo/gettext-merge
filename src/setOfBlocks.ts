import { Block } from './block'
import { hashCompare } from './utils'
import { GetTextTranslation, GetTextTranslations } from 'gettext-parser'

export class SetOfBlocks {
	/** An array of Block objects */
	blocks: Block[]
	path?: string

	/**
	 * Constructor for the Block class.
	 *
	 * @param {Block[]} arr - optional array of Block objects
	 * @param path - optional path to the file
	 */
	constructor(arr?: Block[], path?: string) {
		this.blocks = (arr as Block[]) || []
		this.path = path || undefined
	}

	/**
	 * Add a block to the collection, merging if a duplicate exists.
	 *
	 * @param {Block} block - the block to add
	 * @return {void}
	 */
	add(block: Block): void {
		const duplicate = this.getDuplicate(block.hash())

		if (duplicate) {
			duplicate.merge(block)
		} else {
			this.blocks.push(block)
		}
	}

	/**
	 * Find and return a duplicate block based on the given hash.
	 *
	 * @param {number} hash - the hash to search for
	 * @return {Block | undefined} the duplicate block, if found
	 */
	getDuplicate(hash: number): Block | undefined {
		for (let i = 0; i < this.blocks.length; i++) {
			if (this.blocks[i].hash() === hash) {
				return this.blocks[i]
			}
		}

		return undefined
	}

	/**
	 * Convert the blocks to a string representation.
	 *
	 * @return {string} the string representation of the blocks
	 */
	toStr(): string {
		return this.blocks
			.filter((b) => b.msgid)
			.sort(hashCompare)
			.reduce((prev, curr) => prev + curr.toStr() + '\n\n', '')
	}

	toJson() {
		const newSet: Record<string, { [key: string]: GetTextTranslation }> = {}

		this.blocks
			.filter((b) => b.msgid)
			.sort(hashCompare)
			.forEach((b) => {
				const index = b.msgctxt || ''
				if (!newSet[index]) newSet[index] = {}
				newSet[index][b.msgid || ''] = b.toJson()
			})

		return newSet as GetTextTranslations['translations']
	}

	/**
	 * Adds an array of Block objects to the current instance.
	 *
	 * @param {Block[]} arr - The array of Block objects to add
	 * @return {void}
	 */
	addArray(arr: Block[]): void {
		for (let i = 0; i < arr.length; i++) this.add(arr[i])
		this.blocks.filter((b) => b.msgid)
	}
}
