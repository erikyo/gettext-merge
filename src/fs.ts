import fs from 'fs/promises'
import { SetOfBlocks } from './setOfBlocks.js'
import { Block } from './block.js'
import { hashCompare } from './utils.js'

/**
 * Reads a block of lines from the input array and returns the block along with the remaining lines.
 *
 * @param {string[]} lines - The array of lines to read from.
 * @return {[string[], string[]]} An array containing the block of lines read and the remaining lines.
 */
function readBlock(lines: string[]): [string[], string[]] {
	const block: string[] = []
	let line = lines.pop()

	while (line) {
		block.push(line)
		line = lines.pop()
	}

	return [block, lines]
}

/**
 * Reads and processes blocks from the given array of lines.
 *
 * @param {string[]} lines - Array of lines to process
 * @return {Block[]} The processed blocks
 */
function readBlocks(lines: string[]): Block[] {
	const blocks: Block[] = []

	while (lines.length > 0) {
		const res = readBlock(lines)
		const b = new Block(res[0])
		blocks.push(b)
		lines = res[1]
	}

	return blocks
}

/**
 * Parses the given data string and returns an array of Block objects.
 *
 * @param {string} data - the string to be parsed
 * @return {Block[]} an array of Block objects
 */
export function parseFileContent(data: string): Block[] {
	const lines = data.split(/\r?\n/).reverse()
	const blocks = readBlocks(lines)
	return blocks.sort(hashCompare)
}

/**
 * Writes the consolidated content of the SetOfBlocks to a file specified by the
 * output parameter.
 *
 * @param header - the header to be written
 * @param {SetOfBlocks} blocks - the set of blocks to be consolidated
 * @param {string} output - the path of the file to write the consolidated content
 * @return {Promise<string>} a promise that resolves to the consolidated content
 */
export async function writePo(
	header: Block | undefined,
	blocks: SetOfBlocks,
	output: string
): Promise<string> {
	// add the header
	let consolidated = header ? header.toStr() + '\n\n\n' : ''
	// consolidate the blocks
	consolidated += blocks.cleanup().toStr()

	// TODO: choose whether to override the existing file
	await fs.writeFile(output, consolidated, { encoding: 'utf8', flag: 'w' })
	return consolidated
}
