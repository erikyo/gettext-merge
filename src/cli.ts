#!/usr/bin/env node

import { writePo } from './fs'
import { argv } from './cliArgs'
import { mergePotFile } from './'

/**
 * A function that performs gettext merging.
 *
 * @return {void} no return value
 */
export default function gettextMerger(): void {
	const startTime = new Date()
	const args = argv as unknown as { in: string[]; out: string }
	// Ensure we have exactly two input files
	if (args.in.length <= 1) {
		console.error('You must provide at least two input files.')
		process.exit(1)
	}
	console.log('ðŸ”¥ Merging ' + args.in.join(', ') + ' into ' + args.out)
	mergePotFile(args.in)
		.then(([header, body]) => {
			return writePo(header[0], body, args.out as string)
		})
		.then(() => {
			console.log(
				'ðŸš€ Done in ' + (new Date().getTime() - startTime.getTime()) / 1000 + 's'
			)
			process.exit(0)
		})
		.catch((err) => {
			console.error(err)
		})
}

gettextMerger()

export { gettextMerger }
