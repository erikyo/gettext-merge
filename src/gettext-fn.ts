import fs from 'fs/promises'
import { SetOfBlocks } from './setOfBlocks'
import { Block } from './block'
import { hashCompare } from './utils'

/**
 * Asynchronously runs mergePotStrings with the provided input files and returns
 * the resulting set of blocks.
 *
 * @param {string[]} inputFiles - An array of input file paths to be merged
 * @return {Promise<SetOfBlocks>} A Promise resolving to the set of blocks
 */
export async function runMergePotWithArgs(
	inputFiles: string[]
): Promise<SetOfBlocks> {
	// Ensure we have exactly two input files
	if (inputFiles.length <= 1) {
		console.error('You must provide at least two input files.')
		process.exit(1)
	}

	return await mergePotStrings(inputFiles)
}

/**
 * Reads a block of lines from the input array and returns the block along with the remaining lines.
 *
 * @param {string[]} lines - The array of lines to read from.
 * @return {[string[], string[]]} An array containing the block of lines read and the remaining lines.
 */
function readBlock(lines: string[]): [string[], string[]] {
	const block: string[] = []
	let line = lines.pop()

	while (line) {
		block.push(line)
		line = lines.pop()
	}

	return [block, lines]
}

/**
 * Reads and processes blocks from the given array of lines.
 *
 * @param {string[]} lines - Array of lines to process
 * @return {Block[]} The processed blocks
 */
function readBlocks(lines: string[]): Block[] {
	const blocks: Block[] = []

	while (lines.length > 0) {
		const res = readBlock(lines)
		const b = new Block(res[0])
		blocks.push(b)
		lines = res[1]
	}

	return blocks
}

/**
 * Parses the given data string and returns an array of Block objects.
 *
 * @param {string} data - the string to be parsed
 * @return {Block[]} an array of Block objects
 */
function parseFile(data: string): Block[] {
	const lines = data.split(/\r?\n/).reverse()
	const blocks = readBlocks(lines)
	return blocks.sort(hashCompare)
}

/**
 * Merges multiple arrays of blocks into a single set of blocks.
 *
 * @param {Block[][]} arrays - arrays of blocks to merge
 * @return {SetOfBlocks} a set containing all the blocks from the input arrays
 */
function mergeBlocks(...arrays: Block[][]): SetOfBlocks {
	const set = new SetOfBlocks()
	for (const array of arrays) {
		set.addArray(array)
	}
	return set
}

/**
 * Writes the consolidated content of the SetOfBlocks to a file specified by the
 * output parameter.
 *
 * @param header - the header to be written
 * @param {SetOfBlocks} blocks - the set of blocks to be consolidated
 * @param {string} output - the path of the file to write the consolidated content
 * @return {Promise<string>} a promise that resolves to the consolidated content
 */
export async function writePo(
	header: string,
	blocks: SetOfBlocks,
	output: string
): Promise<string> {
	const consolidated = header + '\n\n\n' + blocks.toStr()

	// override by default
	await fs.writeFile(output, consolidated, { encoding: 'utf8', flag: 'w' })
	return consolidated
}

/**
 * Writes the consolidated content of blocks to a specified output file.
 *
 * @return {Promise<string>} the consolidated content as a string
 * @param filePaths
 */
export async function mergePotStrings(
	filePaths: string[]
): Promise<SetOfBlocks> {
	const mergedSet = await Promise.all(
		filePaths.map(async (filePath) => {
			const fileContent = await fs.readFile(filePath, 'utf8')
			return new SetOfBlocks(parseFile(fileContent)).blocks
		})
	)

	// Retrieve the current blocks from the mergedSet
	const currentBlocks = Array.from(mergedSet)
	// Merge current blocks with the next array of blocks
	return mergeBlocks(...currentBlocks)
}
