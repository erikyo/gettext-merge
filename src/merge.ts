import fs from 'fs/promises'
import { SetOfBlocks } from './setOfBlocks'
import { extractPotHeader, parseFile } from './gettext-fn'
import { Block } from './block'
import { GetTextComment } from './types'

/**
 * Merges multiple arrays of blocks into a single set of blocks.
 *
 * @param {Block[][]} arrays - arrays of blocks to merge
 * @return {SetOfBlocks} a set containing all the blocks from the input arrays
 */
export function mergeBlocks(...arrays: Block[][]): SetOfBlocks {
	const set = new SetOfBlocks()
	for (const array of arrays) {
		set.addArray(array)
	}
	return set
}

/**
 * Asynchronously runs mergePotFile with the provided input files and returns
 * the resulting set of blocks.
 *
 * @param {string[]} inputFiles - An array of input file paths to be merged
 * @return {Promise<SetOfBlocks>} A Promise resolving to the set of blocks
 */
export async function runMergePot(inputFiles: string[]): Promise<[Block[], SetOfBlocks]> {
	// Ensure we have exactly two input files
	if (inputFiles.length <= 1) {
		console.error('You must provide at least two input files.')
		process.exit(1)
	}

	return await mergePotFile(inputFiles)
}

/**
 * Writes the consolidated content of blocks to a specified output file.
 *
 * @return {Promise<string>} the consolidated content as a string
 * @param filePaths
 */
export async function mergePotFile(filePaths: string[]): Promise<[Block[], SetOfBlocks]> {
	const headers: Block[] = []
	const mergedSet = await Promise.all(
		filePaths.map(async (filePath) => {
			const fileContent = await fs.readFile(filePath, 'utf8')
			const [header, content] = extractPotHeader(fileContent)
			// Store the header in the header array
			if (header) headers.push(header)
			// Parse the content and return the SetOfBlocks
			return new SetOfBlocks(parseFile(content)).blocks
		})
	)

	// Retrieve the current blocks from the mergedSet
	const currentBlocks = Array.from(mergedSet)

	// Merge current blocks with the next array of blocks
	return [Array.from(headers), mergeBlocks(...currentBlocks)]
}

/**
 * Merges the contents of multiple POT files into a single string.
 *
 * @param {string[]} fileContents - an array of file contents to be merged
 * @return {string} the merged file contents as a single string
 */
export function mergePotFileContent(fileContents: string[]): string {
	let response: string = ''

	// merge the files
	const mergedSet = fileContents.map((content) => {
		return new SetOfBlocks(parseFile(content)).blocks
	})

	// Retrieve the current blocks from the mergedSet
	const currentBlocks = Array.from(mergedSet)
	// Merge current blocks with the next array of blocks
	response += mergeBlocks(...currentBlocks).toStr()

	return response
}

/**
 * Merge PotObject with the given translationObject.
 *
 * @param {SetOfBlocks[]} translationObject - Array of SetOfBlocks objects to merge
 * @return {SetOfBlocks} Merged SetOfBlocks object
 */
export function mergePotObject(translationObject: SetOfBlocks[]): SetOfBlocks {
	// Merge the SetOfBlocks objects
	const mergedSet = translationObject.map((content) => {
		return new SetOfBlocks(content.blocks).blocks
	})

	return mergeBlocks(...mergedSet)
}

/**
 * Helper function to merge two strings into a unique array, excluding
 * undefined values.
 *
 * @param {string | undefined} current - The current string to merge.
 * @param {string | undefined} other - The other string to merge.
 * @return {string[]} The merged array with unique values.
 */
function mergeUnique(
	current: string | string[] = [],
	other: string | string[] = []
): string[] {
	const mergeSet = new Set([...current, ...other])
	return Array.from(mergeSet)
}

/**
 * Merges two comments into a single comment object.
 *
 * @param {GetTextComment | undefined} current - The current comment object
 * @param {GetTextComment | undefined} other - The other comment object to merge
 * @return {GetTextComment} The merged comment object
 */
export function mergeComments(
	current: GetTextComment | undefined,
	other: GetTextComment | undefined
): GetTextComment {
	return {
		translator: mergeUnique(current?.translator, other?.translator),
		reference: mergeUnique(current?.reference, other?.reference),
		extracted: mergeUnique(current?.extracted, other?.extracted),
		flag: mergeUnique(current?.flag, other?.flag).join('\n'),
		previous: mergeUnique(current?.previous, other?.previous),
	}
}
