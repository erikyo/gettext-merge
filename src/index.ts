#!/usr/bin/env node

import { writePo } from './gettext-fn'
import { argv } from './cliArgs'
import { extractPotHeader } from './utils'
import yargs from 'yargs'
import { readFileSync } from 'fs'
import { runMergePotWithArgs } from './merge'

/**
 * A function that performs gettext merging.
 *
 * @return {void} no return value
 */
export default function gettextMerger() {
	const startTime = new Date()
	const args = argv as unknown as { in: string[]; out: string }
	console.log('ðŸ”¥ Merging ' + args.in.join(', ') + ' into ' + args.out)
	runMergePotWithArgs(args.in)
		.then((res) => {
			// write the header
			// read the file
			const potFile_0 = readFileSync(args.in[0], 'utf8')
			const header = extractPotHeader(potFile_0 as string)
			return writePo(header, res, args.out as string)
		})
		.then(() => {
			console.log(
				'ðŸš€ Done in ' +
					(new Date().getTime() - startTime.getTime()) / 1000 +
					's'
			)
			process.exit(0)
		})
		.catch((err) => {
			console.error(err)
		})
	yargs.help()
}

gettextMerger()

export { gettextMerger }
