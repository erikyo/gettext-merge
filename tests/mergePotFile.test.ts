// Import jest for mocking
// Import the mergePotFile function
import { describe, expect, it } from 'vitest'
import { mergePotFile } from '../src'

describe('Merge Pot Strings Function', () => {
	it('should merge two pot strings correctly', async () => {
		// Call the mergePotFile function
		const result = await mergePotFile([
			'./tests/fixtures/file1.pot',
			'./tests/fixtures/file2.pot',
		])

		// Assert the result
		expect(result[1]).toMatchSnapshot()
	})
})
