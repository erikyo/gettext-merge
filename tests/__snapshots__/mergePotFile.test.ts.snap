// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Merge Pot Strings Function > should merge two pot strings correctly 1`] = `
SetOfBlocks {
  "blocks": [
    Block {
      "comments": {
        "extracted": [
          "Plugin Name of the plugin",
        ],
        "flag": undefined,
        "previous": undefined,
        "reference": undefined,
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "Unicorn Plugin",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "extracted": undefined,
        "flag": undefined,
        "previous": undefined,
        "reference": [
          "admin/Unicorn_Admin_Customizations.php:69",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "Magical Settings",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "extracted": [],
        "flag": "",
        "previous": [],
        "reference": [
          "admin/Unicorn_Admin_Core.php:131",
          "includes/admin/meta-boxes/views/unicorn-product-data.php:5",
        ],
        "translator": [],
      },
      "msgctxt": undefined,
      "msgid": "Unicorn magic updated with sparkles!",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "extracted": undefined,
        "flag": undefined,
        "previous": undefined,
        "reference": [
          "admin/Unicorn_Admin_Customizations.php:94",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "Magical Spells before Summoning",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "extracted": undefined,
        "flag": undefined,
        "previous": undefined,
        "reference": [
          "admin/Unicorn_Admin_Customizations.php:77",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "Summon unicorns automatically",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "extracted": undefined,
        "flag": undefined,
        "previous": undefined,
        "reference": [
          "admin/Unicorn_Admin_Core.php:224",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "Realm of Unicorns",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "extracted": undefined,
        "flag": undefined,
        "previous": undefined,
        "reference": [
          "admin/Unicorn_Admin_Customizations.php:129",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "Enable magical extra checks",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "extracted": undefined,
        "flag": undefined,
        "previous": undefined,
        "reference": [
          "admin/Unicorn_Admin_Core.php:82",
          "admin/Unicorn_Admin_Core.php:83",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "Unicorn Magic",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "extracted": undefined,
        "flag": undefined,
        "previous": undefined,
        "reference": [
          "admin/Unicorn_Admin_Customizations.php:85",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "Mystical Unicorn Appearance",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "extracted": undefined,
        "flag": undefined,
        "previous": undefined,
        "reference": [
          "admin/Unicorn_Admin_Customizations.php:138",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "Append hidden spells on submit",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "extracted": undefined,
        "flag": undefined,
        "previous": undefined,
        "reference": [
          "admin/Unicorn_Admin_Core.php:198",
          "core/Unicorn_Mystical.php:447",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "Are you ready for magic?",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "extracted": [
          "Author of the plugin",
        ],
        "flag": undefined,
        "previous": undefined,
        "reference": undefined,
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "FantasyTech",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "extracted": undefined,
        "flag": undefined,
        "previous": undefined,
        "reference": [
          "admin/Unicorn_Admin_Customizations.php:45",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "Enchanters only",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "extracted": undefined,
        "flag": undefined,
        "previous": undefined,
        "reference": [
          "admin/Unicorn_Admin_Customizations.php:103",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "Automatic Unsummoning",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "extracted": undefined,
        "flag": undefined,
        "previous": undefined,
        "reference": [
          "admin/Unicorn_Admin_Core.php:99",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "Unicorn Settings",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "extracted": undefined,
        "flag": undefined,
        "previous": undefined,
        "reference": [
          "admin/Unicorn_Admin_Core.php:378",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "You've entered the realm of unicorns! Expect enchanting experiences.",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "extracted": undefined,
        "flag": undefined,
        "previous": undefined,
        "reference": [
          "admin/Unicorn_Admin_Core.php:266",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "Mystical Powers",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "extracted": undefined,
        "flag": undefined,
        "previous": undefined,
        "reference": [
          "admin/Unicorn_Admin_Customizations.php:112",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "Horn Enchanting",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "extracted": undefined,
        "flag": undefined,
        "previous": undefined,
        "reference": [
          "admin/Unicorn_Admin_Core.php:329",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "Unicorn Setup",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "extracted": undefined,
        "flag": undefined,
        "previous": undefined,
        "reference": [
          "admin/Unicorn_Admin_Customizations.php:120",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "Enable anti-muggle checks",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "extracted": [
          "Description of the plugin",
        ],
        "flag": undefined,
        "previous": undefined,
        "reference": undefined,
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "A magical plugin that brings unicorns to your website. Transform your content with enchanting features!",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "extracted": [],
        "flag": "",
        "previous": [],
        "reference": [],
        "translator": [
          "Copyright (C) 2024 FantasyTech",
          "This file is distributed under the same license as the Unicorn Plugin.",
          "Copyright (C) 2024 WastedSpace",
          "This file is distributed under the same license as the Useless Plugin.",
        ],
      },
      "msgctxt": undefined,
      "msgid": "",
      "msgid_plural": undefined,
      "msgstr": [
        "",
        "Project-Id-Version: Unicorn Plugin 1.0.0\\n",
        "Report-Msgid-Bugs-To: https://fantasytech.com/support/plugin/unicorn\\n",
        "Last-Translator: FULL NAME <EMAIL@ADDRESS>\\n",
        "Language-Team: LANGUAGE <LL@li.org>\\n",
        "MIME-Version: 1.0\\n",
        "Content-Type: text/plain; charset=UTF-8\\n",
        "Content-Transfer-Encoding: 8bit\\n",
        "POT-Creation-Date: 2024-01-23T12:00:00+00:00\\n",
        "PO-Revision-Date: YEAR-MO-DA HO:MI+ZONE\\n",
        "X-Generator: FantasyCLI 1.0.0-alpha-unicorn\\n",
        "X-Domain: unicorn-plugin\\n",
      ],
    },
    Block {
      "comments": {
        "extracted": [
          "Context: Product description",
        ],
        "flag": undefined,
        "previous": undefined,
        "reference": [
          "includes/admin/meta-boxes/views/html-product-data-description.php:5",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "Product Description",
      "msgid_plural": undefined,
      "msgstr": [
        "Field: Useless Product Description",
      ],
    },
    Block {
      "comments": {
        "extracted": [
          "Context: Featured products widget title",
        ],
        "flag": undefined,
        "previous": undefined,
        "reference": [
          "includes/widgets/class-wc-widget-featured-products.php:40",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "Featured Products",
      "msgid_plural": undefined,
      "msgstr": [
        "Widget Title: Useless Featured Products",
      ],
    },
    Block {
      "comments": {
        "extracted": [
          "Context: Baby games category",
        ],
        "flag": undefined,
        "previous": undefined,
        "reference": [
          "includes/admin/meta-boxes/views/html-product-data-category.php:15",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "Baby-Friendly",
      "msgid_plural": undefined,
      "msgstr": [
        "Field: Useless Baby-Friendly",
      ],
    },
    Block {
      "comments": {
        "extracted": [
          "Chunk file for editing product page",
        ],
        "flag": "",
        "previous": [],
        "reference": [
          "assets/client/admin/chunks/edit-product-page.js:1",
        ],
        "translator": [],
      },
      "msgctxt": undefined,
      "msgid": "Save Changes",
      "msgid_plural": undefined,
      "msgstr": [
        "Button: Save Useless Changes",
      ],
    },
    Block {
      "comments": {
        "extracted": [
          "Context: Cart checkout button",
        ],
        "flag": undefined,
        "previous": undefined,
        "reference": [
          "templates/cart/cart.php:50",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "Proceed to Checkout",
      "msgid_plural": undefined,
      "msgstr": [
        "Button: Useless Proceed to Checkout",
      ],
    },
    Block {
      "comments": {
        "extracted": [
          "Field in generic mapping",
        ],
        "flag": "",
        "previous": [],
        "reference": [
          "includes/admin/importers/mappings/generic.php:26",
        ],
        "translator": [],
      },
      "msgctxt": undefined,
      "msgid": "Menu order",
      "msgid_plural": undefined,
      "msgstr": [
        "Field: Useless Menu Order",
      ],
    },
    Block {
      "comments": {
        "extracted": [
          "Field in products widget class",
        ],
        "flag": "",
        "previous": [],
        "reference": [
          "includes/widgets/class-wc-widget-products.php:53",
        ],
        "translator": [],
      },
      "msgctxt": undefined,
      "msgid": "Widget Title",
      "msgid_plural": undefined,
      "msgstr": [
        "Field: Useless Widget Title",
      ],
    },
    Block {
      "comments": {
        "extracted": [
          "Importer view for successful product update",
        ],
        "flag": "",
        "previous": [],
        "reference": [
          "includes/admin/importers/views/html-csv-import-done.php:28",
        ],
        "translator": [],
      },
      "msgctxt": undefined,
      "msgid": "%s product updated",
      "msgid_plural": "%s products updated",
      "msgstr": [
        "Useless: One product has been updated",
        "Useless: %s products have been updated",
      ],
    },
    Block {
      "comments": {
        "extracted": [
          "Context: Video games category",
        ],
        "flag": undefined,
        "previous": undefined,
        "reference": [
          "includes/admin/meta-boxes/views/html-product-data-category.php:20",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "Game Genre",
      "msgid_plural": undefined,
      "msgstr": [
        "Field: Useless Game Genre",
      ],
    },
    Block {
      "comments": {
        "extracted": [
          "Context: Toy category",
        ],
        "flag": undefined,
        "previous": undefined,
        "reference": [
          "includes/admin/meta-boxes/views/html-product-data-category.php:10",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "Age Group",
      "msgid_plural": undefined,
      "msgstr": [
        "Field: Useless Age Group",
      ],
    },
    Block {
      "comments": {
        "extracted": [
          "Field in advanced product data view",
        ],
        "flag": "",
        "previous": [],
        "reference": [
          "includes/admin/meta-boxes/views/html-product-data-advanced.php:28",
        ],
        "translator": [],
      },
      "msgctxt": undefined,
      "msgid": "Product Status",
      "msgid_plural": undefined,
      "msgstr": [
        "Field: Useless Product Status",
      ],
    },
    Block {
      "comments": {
        "extracted": [
          "Importer view for successful product import",
        ],
        "flag": "",
        "previous": [],
        "reference": [
          "includes/admin/importers/views/html-csv-import-done.php:20",
        ],
        "translator": [],
      },
      "msgctxt": undefined,
      "msgid": "%s product imported",
      "msgid_plural": "%s products imported",
      "msgstr": [
        "Useless: One product has been imported",
        "Useless: %s products have been imported",
      ],
    },
  ],
  "path": undefined,
}
`;
